src/rp2_common/hardware_rtc/rtc.c

```diff
@@ -131,43 +144,51 @@ static void rtc_irq_handler(void) {
+bool rtc_set_alarm(const datetime_t *t, rtc_callback_t user_callback) {
     rtc_disable_alarm();
 
-    // Only add to setup if it isn't -1
-    rtc_hw->irq_setup_0 = ((t->year  < 0) ? 0 : (((uint)t->year)  << RTC_IRQ_SETUP_0_YEAR_LSB )) |
-                          ((t->month < 0) ? 0 : (((uint)t->month) << RTC_IRQ_SETUP_0_MONTH_LSB)) |
-                          ((t->day   < 0) ? 0 : (((uint)t->day)   << RTC_IRQ_SETUP_0_DAY_LSB  ));
-    rtc_hw->irq_setup_1 = ((t->dotw  < 0) ? 0 : (((uint)t->dotw)  << RTC_IRQ_SETUP_1_DOTW_LSB)) |
-                          ((t->hour  < 0) ? 0 : (((uint)t->hour)  << RTC_IRQ_SETUP_1_HOUR_LSB)) |
-                          ((t->min   < 0) ? 0 : (((uint)t->min)   << RTC_IRQ_SETUP_1_MIN_LSB )) |
-                          ((t->sec   < 0) ? 0 : (((uint)t->sec)   << RTC_IRQ_SETUP_1_SEC_LSB ));
-
-    // Set the match enable bits for things we care about
-    if (t->year  >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_YEAR_ENA_BITS);
-    if (t->month >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_MONTH_ENA_BITS);
-    if (t->day   >= 0) hw_set_bits(&rtc_hw->irq_setup_0, RTC_IRQ_SETUP_0_DAY_ENA_BITS);
-    if (t->dotw  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_DOTW_ENA_BITS);
-    if (t->hour  >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_HOUR_ENA_BITS);
-    if (t->min   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_MIN_ENA_BITS);
-    if (t->sec   >= 0) hw_set_bits(&rtc_hw->irq_setup_1, RTC_IRQ_SETUP_1_SEC_ENA_BITS);
+    uint32_t s0 = 0, s1 = 0;
 
     // Does it repeat? I.e. do we not match on any of the bits
     _alarm_repeats = rtc_alarm_repeats(t);
 
+    if( (!valid_datetime(t) && !_alarm_repeats) || !user_callback)	// none of the parameters is valid
+        return false;
+
+    // Set the match enable bits for things we care about
+    if (t->year  >= 0)
+        s0 |= RTC_IRQ_SETUP_0_YEAR_ENA_BITS  | (((uint)t->year)  << RTC_IRQ_SETUP_0_YEAR_LSB );
+    if (t->month >= 1)
+        s0 |= RTC_IRQ_SETUP_0_MONTH_ENA_BITS | (((uint)t->month) << RTC_IRQ_SETUP_0_MONTH_LSB);
+    if (t->day   >= 1)
+        s0 |= RTC_IRQ_SETUP_0_DAY_ENA_BITS   | (((uint)t->day)   << RTC_IRQ_SETUP_0_DAY_LSB  );
+    if (t->dotw  >= 0)
+        s1 |= RTC_IRQ_SETUP_1_DOTW_ENA_BITS  | (((uint)t->dotw)  << RTC_IRQ_SETUP_1_DOTW_LSB);
+    if (t->hour  >= 0)
+       s1 |= RTC_IRQ_SETUP_1_HOUR_ENA_BITS   | (((uint)t->hour)  << RTC_IRQ_SETUP_1_HOUR_LSB);
+    if (t->min   >= 0)
+        s1 |= RTC_IRQ_SETUP_1_MIN_ENA_BITS   | (((uint)t->min)   << RTC_IRQ_SETUP_1_MIN_LSB);
+    if (t->sec   >= 0)
+        s1 |= RTC_IRQ_SETUP_1_SEC_ENA_BITS   | (((uint)t->sec)   << RTC_IRQ_SETUP_1_SEC_LSB);
+    else if (_alarm_repeats == -1) // repeatable every second! All entries are -1
+    {
+        datetime_t tNew;
+        rtc_get_datetime(&tNew);
+        s1 = RTC_IRQ_SETUP_1_SEC_ENA_BITS | ((((uint)tNew.sec + 1) % 60) << RTC_IRQ_SETUP_1_SEC_LSB);
```

Comment:
This seems race prone; is using the RTC clock for an every second interrupt useful?

Reply:
at least it works fine - updating my display every second while pushing the current time on UART.
But there is another more critical thing missing. ATM an attached exclusive interrupt can't be removed.

Reply:
the current ISR still enables rtc_alarm before user callback has returned. Maybe we should change this too?

Reply:
Another question, usually RTC uses 32.768kHz clock, here 46.875kHz is programmed. What is the cause for this?

Reply:
The RP2040 datasheet says: "XOSC @ 12MHz / 256 = 46875Hz. To get a 1Hz reference CLKDIV_M1 should be set to 46874."
As long as `CLKDIV_M1` is set to an appropriate value, I don't think it matters what frequency the RTC clock runs at? (Looks like https://github.com/raspberrypi/pico-sdk/blob/develop/src/rp2_common/hardware_clocks/clocks.c#L187 is where clk_rtc is actually configured)

Reply:
Exactly,  12MHz / 256...  I found it. Would be nice if the comment you linked would say the same